// @kiltprotocol/type-definitions/lib/cjs/types_18.js
"use strict";
/**
 * Copyright (c) 2018-2023, BOTLabs GmbH.
 *
 * This source code is licensed under the BSD 4-Clause "Original" license
 * found in the LICENSE file in the root directory of this source tree.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.types18 = void 0;
const mergeType_js_1 = require("./mergeType.js");
const types_17_js_1 = require("./types_17.js");
exports.types18 = (0, mergeType_js_1.mergeType)(
// Use the old types as the base of the new types.
types_17_js_1.types17, 
// We add these new type:
{
    // DID management update
    DidCreationDetails: {
        did: 'DidIdentifierOf',
        newKeyAgreementKeys: 'BTreeSet<DidEncryptionKey>',
        newAttestationKey: 'Option<DidVerificationKey>',
        newDelegationKey: 'Option<DidVerificationKey>',
        newEndpointUrl: 'Option<Url>',
    },
    DidUpdateDetails: {
        newAuthenticationKey: 'Option<DidVerificationKey>',
        newKeyAgreementKeys: 'BTreeSet<DidEncryptionKey>',
        attestationKeyUpdate: 'DidVerificationKeyUpdateAction',
        delegationKeyUpdate: 'DidVerificationKeyUpdateAction',
        publicKeysToRemove: 'BTreeSet<KeyIdOf>',
        newEndpointUrl: 'Option<Url>',
    },
}, 
// Remove old DID types
['DidCreationOperation', 'DidUpdateOperation', 'DidDeletionOperation']);
